src/core/services/shift_service.py

```diff
@@ -59,7 +60,28 @@ async def get_shift(self, id: UUID) -> Shift:
         return await self.__shift_repository.get(id)
 
     async def update_shift(self, id: UUID, update_shift_data: ShiftUpdateRequest) -> Shift:
-        return await self.__shift_repository.update(id, Shift(**update_shift_data.dict(exclude_unset=True)))
+        shift = await self.__shift_repository.get(id)
+        if shift.status in (Shift.Status.CANCELING, Shift.Status.FINISHED):
+            raise HTTPException(
+                status_code=HTTPStatus.FORBIDDEN, detail="Нельзя изменять данные закончившейся или отмененной смены"
+            )
+        update_shift_data = update_shift_data.dict(exclude_unset=True)

```

Comment:
Не надо выполнять преобразование в словарь. С объектами работать намного проще, удобнее и нагляднее.
```suggestion
        shift: Shift = await self.__shift_repository.get(id)
        if shift.status in (Shift.Status.CANCELING, Shift.Status.FINISHED):
            raise ...

        if shift.status == Shift.Status.STARTED:
            if update_shift_data.started_at:
                raise ...

        if update_shift_data.started_at and not update_shift_data.finished_at:
            if update_shift_data.started_at >= shift.finished_at:
                raise ...
            shift.started_at = update_shift_data.started_at

        if not update_shift_data.started_at and update_shift_data.finished_at:
            if shift.started_at >= update_shift_data.finished_at:
                raise ...
            shift.finished_at = update_shift_data.finished_at

        if update_shift_data.title:
            shift.title = update_shift_data.title
        if update_shift_data.final_message:
            shift.final_message = update_shift_data.final_message

        return await self.__shift_repository.update(id, shift)
```

Reply:
Последние проверки и присвоения можно заменить на тернальный оператор, но это дело вкуса: 
        shift.title = update_shift_data.title if update_shift_data.title else shift.title

Reply:
Теперь все поля обязательные (см. коммент Антона ниже), и всё немного по другому:)